### PR TITLE
BUG: Groupby not keeping object dtype when infer string is set

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1904,8 +1904,7 @@ def _agg_py_fallback(
             ser = Series(values, copy=False)
         else:
             # We only get here with values.dtype == object
-            # TODO: special case not needed with ArrayManager
-            df = DataFrame(values.T)
+            df = DataFrame(values.T, dtype=values.dtype)
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:1907]: Perhaps slightly out of scope, but I think it'd be good to add an assert on values.dtype == object here

**Review by rhshadrach — COMMENTED** [review]
Looks good

**Comment by rhshadrach** [comment]
Is this a regression with no related issue?

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (phofl) [reply pandas/core/groupby/groupby.py:1907]: Yeah I'd rather keep this out of here

**Comment by phofl** [comment]
I am working on a build that uses arrow backed strings by default, but this turns up dozens of bugs related to either the new string dtype we added for 2.1 or arrow strings more generally, not opening issues since I am fixing them when I come across them

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @phofl

**Pull request merged**